apps/aecore/src/aec_chain.erl

```diff
@@ -862,9 +868,22 @@ do_write_block(Block, TopHeader, TopBlock, HeadersDb, BlocksDb) ->
                 {ok, SerializedBlock} ->
                     {error, {block_already_stored, SerializedBlock}};
                 {error, not_found} ->
+                    BlockHeight = aec_headers:height(Header),
+                    TopBlockHeight = aec_blocks:height(TopBlock),
+                    {_ok_OnlyForOrderedBlocks, {StateHeight, Trees}} =
+                            aec_state:apply_txs(aec_blocks:txs(Block), BlockHeight),
+
+                    %% If this is out of order block, its fine to store it,
+                    %% but we don't have possibility to compute state - its missing
+                    %% TODO: can we back propagete state when we check for successor in aec_state?
+                    %%       currently updating db may cause data inconsistency or deadlock
+                    BlockMaybeWithTrees = case StateHeight == BlockHeight of
+                                              true -> aec_blocks:set_trees(Block, Trees);
+                                              false -> Block
+                                          end,
                     %% Store block.
                     {ok, SerializedBlock} =
-                        aec_blocks:serialize_for_network(Block),
+                        aec_blocks:serialize_for_network(BlockMaybeWithTrees),
```

Comment:
@sennui Just ensuring we are on the same page:
* The chain service stores block without state trees (see type `blocks_db_value()`). Function `aec_blocks:serialize_for_network` removes state trees - if any - from block;
* This `do_write_block` function returns `NewTopBlock`, then cached in state `#top_state.top_block`, that is without state trees.

So this `BlockMaybeWithTrees` may be deleted. (Not needed - just ensuring I understand correctly the intention.)

Reply:
@lucafavatella 

* serialize for network surly removes trees thats understandable and required

* a block is kept in block record. it has state trees in it (in a record). Don't we cache it?

We can compute trees for a block only if its Prev+1 Block. So this is why I called it Maybe.
It is fine to have block record out of order and without trees.

Reply:
> a block is kept in block record. it has state trees in it (in a record). Don't we cache it?

No, currently we do not cache the state trees in the chain service. And, in order to keep things simpler, I suggest not to for the moment.

More in detail, in the chain service's `#state`, the cached top block `#state.top#top_state.top_block` has type `aec_blocks:block_deserialized_from_network()` i.e. does not have state trees.

Reply:
@lucafavatella OK, let me remove that for now. I added note in a ticket that should evolve to its own ticket - we may consider keeping trees in cached #block